Monorepo changes & Package Manager Internal Docs (#418)
flarum/docs: docs/contributing.md

### Setting Up a Local Codebase

> [flarum/flarum](https://github.com/flarum/flarum) is a "skeleton" application which uses Composer to download the core package and a bunch of extensions within the monorepo [flarum/framework](https://github.com/flarum/framework). In order to work on these, we recommend forking and cloning the repository into a [Composer path repository](https://getcomposer.org/doc/05-repositories.md#path):

Reviewer: ```suggestion
[flarum/flarum](https://github.com/flarum/flarum) is a "skeleton" application which uses Composer to download the core package and a bunch of extensions. Source code for Flarum core, extensions, and all packages used by the aforementioned is located in the Flarum monorepo [flarum/framework](https://github.com/flarum/framework). In order to contribute to these, you'll need to fork and clone the monorepo repository locally, and then add it to your dev environment as a [Composer path repository](https://getcomposer.org/doc/05-repositories.md#path):
```

```bash
git clone https://github.com/flarum/flarum.git
cd flarum

> # Set up a Composer path repository for Flarum monorepo packages
> composer config repositories.0 path "packages/framework/*"

Reviewer: ```suggestion
# Set up a Composer path repository for Flarum monorepo packages
composer config repositories.0 path "PATH_TO_MONOREPO/*/*"
```

Reviewer: Note that the double star (/*/*) is deliberate. Should we mention that?

Author: Oh right, we changed the monorepo structure, I don't believe a mention is necessary, most will copy past.

> composer config repositories.0 path "PATH_TO_MONOREPO/*/*"
> git clone https://github.com/<username>/framework.git packages/framework

Author: I think it's easier to have it as `packages` directly

```suggestion
composer config repositories.0 path "packages/*/*"
git clone https://github.com/<username>/framework.git packages/framework
```

Reviewer: Not sure about that, actually. `packages` could refer to the new monorepo, or to a generic packages directory where all flarum extensions are installed.

Author: I was more referring to the fact that the command right after it clones the monorepo fork to the same packages path, so it's a no brainer copy past for new contributors.

Though I don't feel strongly about this, just a thought I had when I noticed the second command.

Will make them consistent.

Reviewer: Ah but then it doesn't work, as `packages/*/*` would refer to the folders in the root level of `framework`, not the extensions/packages inside.

If we want to be very general, perhaps we should clone to `monorepos/flarum`, and then have the path be `monorepos/*/*/*`?

Author: Oh wait, I missed what you actually meant. True, it wouldn't be correctly pointing to the actual monorepo packages.

perhaps a simple change to `packages/framework/*/*`, even though this wouldn't take into consideration local 3rd party extensions, since it's not the point of the contributing guide I think it should be fine, for extension development we have this https://docs.flarum.org/extend/start#extension-packaging.

For the more general solution, I don't think there'll be many monorepos in the ecosystem, and even if there will be, they probably won't have the same subdirectory structure Flarum has, so not sure it's worth it.

Author: Let's just leave it as `PATH_TO_MONOREPO` 👍🏼, both commands are in sync now.

Flarum's front-end code is written in ES6 and transpiled into JavaScript. During development you will need to recompile the JavaScript using [Node.js](https://nodejs.org/). **Please do not commit the resulting `dist` files when sending PRs**; this is automatically taken care of when changes are merged into the `master` branch.

```bash
cd packages/framework/framework/core/js
npm install
npm run dev
```

The process is the same for extensions.

```bash
cd packages/framework/extensions/tags/js
npm install
npm link ../../../../framework/core/js
npm run dev

> ```

Author: I missed this unnecessary bit. Yarn workspaces for the win, I should go through mentions of `npm` and change them to `yarn` as that's what we use now in the monorepo.

Reviewer: I am very happy with this change :grin:

### PHP

> Flarum follows the [PSR-2](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md) coding standard and the [PSR-4](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-4-autoloader.md) autoloading standard. On top of this, we conform to a number of [other style rules](https://github.com/flarum/framework/blob/master/.styleci.yml). We use PHP 7 type hinting and return type declarations where possible, and [PHPDoc](https://docs.phpdoc.org/) to provide inline documentation. Try and mimic the style used by the rest of the codebase in your contributions.

Reviewer: This should probably use the `main` branch

Author: ```suggestion
Flarum follows the [PSR-2](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md) coding standard and the [PSR-4](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-4-autoloader.md) autoloading standard. On top of this, we conform to a number of [other style rules](https://github.com/flarum/framework/blob/main/.styleci.yml). We use PHP 7 type hinting and return type declarations where possible, and [PHPDoc](https://docs.phpdoc.org/) to provide inline documentation. Try and mimic the style used by the rest of the codebase in your contributions.
```
